text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,802 @@
+    ) -> &mut MessageSender<S, 'a>;
+
+    fn options(
+        &mut self,
+        options: SendOptions,
+    ) -> &mut MessageSender<S, 'a>;
+}
+
+impl<S: SendMessage> MessageSender<S, '_> {
+    fn send(&mut self) -> io::Result<usize>;
+}
+
+impl<S: SendMessageTo> MessageSender<S, '_> {
+    fn send_to(&mut self, addr: &S::SocketAddr) -> io::Result<usize>;
+}
+```
+
+## Receiving messages
+
+### The `RecvMessage` and `RecvMessageFrom` traits
+
+These traits are implemented for socket types that implement `recvmsg`. They
+are equivalent to the current `recv_vectored_with_ancillary` and
+`recv_vectored_with_ancillary_from` functions, but also accept `RecvOptions`
+for per-recv flags, and aren't specific to `AF_UNIX` sockets.
+
+```rust
+trait RecvMessage {
+    fn recv_message(
+        &self,
+        bufs: &mut [IoSliceMut<'_>],
+        ancillary_data: &mut AncillaryData<'_>,
+        options: RecvOptions,
+    ) -> io::Result<(usize, RecvResult)>;
+}
+
+trait RecvMessageFrom {
+    type SocketAddr;
+
+    fn recv_message_from(
+        &self,
+        bufs: &mut [IoSliceMut<'_>],
+        ancillary_data: &mut AncillaryData<'_>,
+        options: RecvOptions,
+    ) -> io::Result<(usize, RecvResult, Self::SocketAddr)>;
+}
+```
+
+The `RecvMessage` trait will be implemented for:
+- `std::net::TcpStream`
+- `std::net::UdpSocket`
+- `std::os::unix::net::UnixDatagram`
+- `std::os::unix::net::UnixStream`
+
+The `RecvMessageFrom` trait will be implemented for:
+- `std::net::UdpSocket`
+- `std::os::unix::net::UnixDatagram`
+
+These traits aren't sealed, so they can be implemented by third-party libraries.
+
+### `RecvOptions`
+
+The `std::os::unix::net::RecvOptions` type wraps flags passed to the Unix
+sockets API functions [`recv`] and [`recvmsg`]. It has inherent methods for
+options defined by the POSIX standard.
+
+The `MSG_CMSG_CLOEXEC` flag is always set on platforms for which it is defined.
+
+[`recv`]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/recv.html
+[`recvmsg`]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/recvmsg.html
+
+```rust
+struct RecvOptions;
+
+impl Default for RecvOptions;
+
+impl RecvOptions {
+    fn new() -> RecvOptions;
+
+    fn as_recv_flags(&self) -> c_int;
+
+    fn custom_flags(&mut self, flags: c_int) -> &mut Self;
+
+    // MSG_OOB
+    fn out_of_band(&mut self, out_of_band: bool) -> &mut Self;
+
+    // MSG_PEEK
+    fn peek(&mut self, peek: bool) -> &mut Self;
+
+    // MSG_WAITALL
+    fn wait_all(&mut self, wait_all: bool) -> &mut Self;
+}
+```
+
+### `RecvResult`
+
+The `std::os::unix::net::RecvResult` type wraps flags returned from the Unix
+sockets API function [`recvmsg`]. These flags indicate various status
+conditions about the received data. It has inherent methods for flags defined
+by the POSIX standard.
+
+```rust
+struct RecvResult;
+
+impl RecvResult {
+    fn new(flags: c_int) -> RecvResult;
+
+    fn custom_flags(&self, flags: c_int) -> bool;
```

Comment:
This should be `fn custom_flags(&self, flag_mask: c_int) -> c_int;`

Reply:
Hmm, I had expected the implementation would be something like this, which would (I think?) work for multi-bit flags:

```rust
fn custom_flags(&self, flags: c_int) -> bool {
    self.bits & flags == flags
}
```

Is there a case that would need the user to inspect the masked bits directly? I'm fine with using that API, but would like to make sure I understand it so I can write good examples in the docs.

Reply:
what happens when you have a 2-bit field and you want to see if it's 2? `self.bits & flags == flags` can't do that because the mask would need to be e.g. `0x30` but would need to be compared against `0x20`

Reply:
I'm not sure what that means, sorry. If the platform has `msg_flags` flags like this:

```rust
// set on `(*struct msghdr)->msg_flags
const MSG_DID_MAGIC: c_int = 0x30;
```

Then testing whether that flag was set would need to `flags & MSG_DID_MAGIC == MSG_DID_MAGIC`.

Did you have a particular platform / flag in mind? Looking up its definition in the OS headers or `libc` might be helpful.

Reply:
if you had:
```rust
const MSG_FOO_MODE_MASK: c_int = 0x30;
const MSG_FOO_MODE_A: c_int = 0;
const MSG_FOO_MODE_B: c_int = 0x10;
const MSG_FOO_MODE_C: c_int = 0x20;
const MSG_FOO_MODE_D: c_int = 0x30;

fn foo_mode_is(msg_flags: c_int, mode: c_int) -> bool {
    msg_flags & MSG_FOO_MODE_MASK == mode
}
```
then we should design it so you can have an equivalent for `foo_mode_is(msg_flags, MSG_FOO_MODE_C)` where it checks that both bit `1 << 4` is zero and bit `1 << 5` is one, since both bits need to be checked to match and not all bits are ones. `flags & mask == mask` is only capable of checking for bits that are ones, it is incapable of checking for zeros.

Reply:
That's the first time I've seen that sort of structure in a `msghdr` flag. What platform is that describing?

Reply:
Also, if there are any OSes with flags like that, then I'd probably try to have an API like this (names illustrative):

```
impl RecvResult {
fn custom_flags(&self, flags: c_int) -> bool;
fn as_raw_flags(&self) -> c_int;
```

So that most platform-specific flags can be checked easily, and any that use multi-bit values with masking can do that part in their own code.

Reply:
> That's the first time I've seen that sort of structure in a `msghdr` flag. Which OS / architecture has recvmsg message flags like that?

nothing I'm aware of, I was just attempting to elaborate on @riking's comment. That said, multi-bit flags like that could totally exist or be added sometime in the future.